src/main/java/org/ibp/api/java/impl/middleware/observationunits/ObservationUnitImportRequestValidator.java

```diff
@@ -0,0 +1,165 @@
+				.map(ObservationUnitImportRequestDto::getGermplasmDbId).collect(Collectors.toList());
+		final GermplasmSearchRequestDto germplasmSearchRequestDto = new GermplasmSearchRequestDto();
+		germplasmSearchRequestDto.setGermplasmDbIds(germplasmDbIds);
+		final Map<String, GermplasmDTO> germplasmDTOMap = this.germplasmService.searchFilteredGermplasm(germplasmSearchRequestDto, null)
+			.stream().collect(Collectors.toMap(GermplasmDTO::getGermplasmDbId, Function.identity()));
+
+		final StudySearchFilter studySearchFilter = new StudySearchFilter();
+		studySearchFilter.setStudyDbIds(studyDbIds);
+		final Map<String, StudyInstanceDto> studyInstancesMap =
+			this.studyInstanceService.getStudyInstances(studySearchFilter, null).stream()
+				.collect(Collectors.toMap(StudyInstanceDto::getStudyDbId, Function.identity()));
+
+		final Integer index = 1;
+		final Iterator<ObservationUnitImportRequestDto> iterator = observationUnitImportRequestDtos.iterator();
+		while (iterator.hasNext()) {
+			final ObservationUnitImportRequestDto dto = iterator.next();
+			if (StringUtils.isEmpty(dto.getProgramDbId())) {
+				this.errors.reject("observation.unit.import.programDbId.null", new String[] {index.toString()}, "");
+				iterator.remove();
+				continue;
+			}
+
+			if (StringUtils.isEmpty(dto.getTrialDbId())) {
+				this.errors.reject("observation.unit.import.trialDbId.null", new String[] {index.toString()}, "");
+				iterator.remove();
+				continue;
+			}
+
+			if (StringUtils.isEmpty(dto.getStudyDbId())) {
+				this.errors.reject("observation.unit.import.studyDbId.null", new String[] {index.toString()}, "");
+				iterator.remove();
+				continue;
+			}
+
+			if (!studyInstancesMap.containsKey(dto.getStudyDbId())
+				|| !studyInstancesMap.get(dto.getStudyDbId()).getProgramDbId().equalsIgnoreCase(dto.getProgramDbId())
+				|| !studyInstancesMap.get(dto.getStudyDbId()).getTrialDbId().equals(dto.getTrialDbId())) {
+				this.errors.reject("observation.unit.import.no.study", new String[] {index.toString()}, "");
+				iterator.remove();
+				continue;
+			}
+
+			if (StringUtils.isEmpty(dto.getGermplasmDbId())) {
+				this.errors.reject("observation.unit.import.germplasmDbId.null", new String[] {index.toString()}, "");
+				iterator.remove();
+				continue;
+			}
+
+			if (!germplasmDTOMap.containsKey(dto.getGermplasmDbId())) {
+				this.errors.reject("observation.unit.import.germplasmDbId.invalid", new String[] {index.toString()}, "");
+				iterator.remove();
+				continue;
+			}
+
+			final ObservationUnitPosition position = dto.getObservationUnitPosition();
+
+			if (position == null || StringUtils.isEmpty(position.getEntryType())) {
+				this.errors.reject("observation.unit.import.entry.type.required", new String[] {index.toString()}, "");
+				iterator.remove();
+				continue;
+			}
+
+			final List<String> entryTypes =
+				this.ontologyService.getStandardVariable(TermId.ENTRY_TYPE.getId(), dto.getProgramDbId()).getEnumerations()
```

Comment:
This could be an expensive operation. Move this outside of loop. Maybe call this method on unique programDbIds in case there are different programDbId per observationUnit.